core/modules/link/link.module

```diff
@@ -930,6 +945,67 @@ function link_field_process($element, $form_state, $complete_form) {
   return $element;
 }
 
+/**
+ * Link field element validation.
+ */
+function link_field_element_validate(&$element, &$form_state, $complete_form) {
+  $value = $element['#value'];
+  $is_default_field_value = FALSE;
+
+  // Check if this element is used in Field API Link field.
+  if (isset($element['#link_field_options']['field_name'])) {
+    // Do not validate when submitting a field instance settings form.
+    if (empty($element['#entity'])) {
+      return;
+    }
+    else {
+      $instance = $element['#link_field_options']['instance'];
+      $delta = $element['#delta'];
+      $is_default_field_value =  isset($instance['default_value'][$delta]['url']) && $value['url'] === $instance['default_value'][$delta]['url'] && !$instance['required'];
```

Comment:
Here's the section where the field instance is used, intermingling Field API with the generic `link_field` element. If it's possible to pass separate properties here instead of the entire instance, that would fully decouple the Form API element.